Slack notifications for Meetup and WordCamp application changes

vedanshujain opened the pull request:
Pending:
- [x] Decline notification for WordCamp and Meetup
- [x] Schedule notification for WordCamp
- [x] Travis build fix

coreymckrill commented on public_html/wp-content/plugins/wcpt/wcpt-meetup/class-meetup-admin.php:
```diff
@@ -478,6 +480,58 @@ public function maybe_update_organizers( $new_status, $old_status, $post ) {
+			$title           = "New meetup group added";
+
+			$message = sprintf(
+				"Lets welcome new WordPress meetup%s%s, to the chapter! :tada: :community: :wordpress:\n%s",
```
```suggestion
				"Let's welcome the new WordPress meetup group%s%s to the chapter! :tada: :community: :wordpress:\n%s",
```

coreymckrill commented on public_html/wp-content/plugins/wcpt/wcpt-wordcamp/wordcamp-admin.php:
```diff
@@ -706,6 +707,56 @@ public function mark_date_added_to_planning_schedule( $wordcamp ) {
+
+			$message = sprintf(
+				"WordCamp $city has been scheduled for %s. :tada: :community: :wordpress:\n\n%s",
+				date( 'd-M-Y', $start_date ),
```
I think the `'d-M-Y'` date format will be confusing to some people (at least in the US). It might be better to go with something more longform like `January 9th, 2019`.

Also, should we include the end date here too for multiday camps?

coreymckrill replied:
Maybe something like `WordCamp $city has been scheduled for a start date of %s.`

Another thought, what if we linked `WordCamp $city` to the event site?

vedanshujain replied:
I think we should do both. show a URL as well as link it.

coreymckrill reviewed (changes requested):
A few minor suggestions and questions added. Also, I noticed some coding standards stuff that phpcs should catch.

vedanshujain commented on public_html/wp-content/plugins/wcpt/wcpt-event/class-event-application.php:
```diff
@@ -208,4 +230,32 @@ public function notify_applicant_application_received( $email_address, $meetup_c
+	public function notify_new_application_in_slack() {
+
+		// Not translating because this will be sent to community events slack channel.
+		$message = sprintf( " A %s application for %s has been received.", $this->get_event_label(), $this->get_event_location() );
```
nope. good catch, will fix

vedanshujain commented on public_html/wp-content/plugins/wcpt/wcpt-event/notification.php:
```diff
@@ -0,0 +1,141 @@
+ */
+function notify_in_events_channel ( $attachment ) {
+	return wcpt_slack_notify( COMMUNITY_EVENTS_SLACK, $attachment );
+}
```
so I was thinking along the lines of whitelisting only few channels. but now that its in a finished state, I would agree with you that this abstractions is not really necessary. if we want to whitelist, we should do so inside `wcpt_slack_notify`

vedanshujain commented on public_html/wp-content/plugins/wcpt/wcpt-event/notification.php:
```diff
@@ -0,0 +1,141 @@
+
+	$status_change_logs = get_post_meta( $event_id, '_status_change' ) ?? array();
+
+	$user_ids = array_merge( $user_ids, wp_list_pluck( $status_change_logs, 'user_id' ) );
```
so I am initializing with empty array above `?? array();` in case `_status_change` log is not present. wp_list_pluck should be fine with empty array I think? is that not the case

coreymckrill replied:
Here's the basic logic in `wp_list_pluck`:

```php
foreach ( $this->output as $key => $value ) {
	if ( is_object( $value ) ) {
		$this->output[ $key ] = $value->$field;
	} else {
		$this->output[ $key ] = $value[ $field ];
	}
}
```

So it looks like if `$value[ $field ]` doesn't exist, it would throw the `undefined index` warning. But I haven't tested it.

vedanshujain replied:
oh yes, if the field is not present there would be notice. but given that status log is created by our code, it will always have those fields. keeping that in mind there should be a notice if the field is not present.
otherwise i cannot use `wp_list_pluck` and have to use a loop to get all the elements.

coreymckrill replied:
> given that status log is created by our code, it will always have those fields

Right, but if there aren't any status log entries yet, so the default empty array is returned via `?? array();`, would that throw a notice? I guess maybe the foreach loop wouldn't even get far enough to detect an undefined index if the array has no items...

vedanshujain replied:
looks like default empty array will not throw a notice. and you are correct that foreach loop wouldn't event get far enough to detect an undefined index if the array has not items.

vedanshujain commented on public_html/wp-content/plugins/wcpt/wcpt-meetup/class-meetup-admin.php:
```diff
@@ -478,6 +480,58 @@ public function maybe_update_organizers( $new_status, $old_status, $post ) {
+			}
+
+			if ( 'wcpt-mtp-active' === $new_status ) {
+				return $this->notify_new_meetup_group_in_slack( $meetup );
```
so both of these approaches works. but I agree that `self::` feels more correct. will change that.
its static so that we can easily call these from cli if need be.

vedanshujain committed:
WCPT: Send notifications to community-events or community-team slack channel on status changes in WordCamp/Meetup applications

Whenever a WordCamp/Meetup application is received, accepted or declined, we will send out notifications. This will help in surfacing interesting changes happening in the community.
This is also discussed in https://make.wordpress.org/community/2018/12/24/slack-notifications-for-wordcamp-and-meetup-application-updates/

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8083 74240141-8908-4e6f-9713-ba540dce6ec7
